cmdstanpy/compiler_opts.py

```diff
@@ -149,10 +156,37 @@ def validate_cpp_opts(self) -> None:
                 val = self._cpp_options[key]
                 if not isinstance(val, int) or val < 0:
                     raise ValueError(
-                        '{} must be a non-negative integer value,'
-                        ' found {}.'.format(key, val)
+                        f'{key} must be a non-negative integer value,'
+                        f' found {val}.'
                     )
 
+    def validate_user_header(self) -> None:
+        """
+        User header exists.
+        Raise ValueError if bad config is found.
+        """
+        if self._user_header != "":
+            if not (
+                os.path.exists(self._user_header)
+                and os.path.isfile(self._user_header)
+            ):
+                raise ValueError(
+                    f"User header file {self._user_header} cannot be found"
+                )
+            if self._user_header[-4:] != '.hpp':
+                raise ValueError(
+                    f"Header file must end in .hpp, got {self._user_header}"
+                )
+            if "allow_undefined" not in self._stanc_options:
+                self._stanc_options["allow_undefined"] = True
+            # set full path
+            self._user_header = os.path.abspath(self._user_header)
+
+            if ' ' in self._user_header:
```

Comment:
is it possible to use TemporaryCopiedFile here?

Reply:
I couldn't think of any way to do so inside CompilerOpts, because the context manager would get destroyed when the function returned. We could use one in the `compile` call, but you end up with a pretty ugly 'if we have a file, use `TemporaryCopiedFile` and do all this stuff, else, do all the exact same stuff' with no real way of de-duplicating it. It also starts to pollute the logic of `compile()` to do some argument validation again

Happy to take suggestions if you can see a way to do it. I figure this way of doing it is safe and simple and we can improve it in the future if need be

Reply:
this is an expert-level feature, correct?  in which case, this is OK.